pump/storage/storage.go

```diff
@@ -686,96 +685,95 @@ func (a *Append) doGCTS(ts int64) {
 		wg.Done()
 	}()
 
-	batch := new(leveldb.Batch)
 	l0Trigger := defaultStorageKVConfig.CompactionL0Trigger
 	if a.options.KVConfig != nil && a.options.KVConfig.CompactionL0Trigger > 0 {
 		l0Trigger = a.options.KVConfig.CompactionL0Trigger
 	}
 
+	getL0Num := func() (int, error) {
+		nStr, err := a.metadata.GetProperty("leveldb.num-files-at-level0")
+		if err != nil {
+			return 0, errors.Annotate(err, "get `leveldb.num-files-at-level0` property failed")
+		}
+
+		l0Num, err := strconv.Atoi(nStr)
+		if err != nil {
+			return 0, errors.Annotatef(err, "parse `leveldb.num-files-at-level0` result to int failed, str: %s", nStr)
+		}
+		return l0Num, nil
+	}
+
+	var alreadyGcTS int64
+	var finished bool
+	var err error
+	for {
+		alreadyGcTS, finished, err = a.batchGC(alreadyGcTS, ts, 100, l0Trigger, getL0Num)
+		if err != nil {
+			log.Error("batch GC failed", zap.Error(err))
+			break
+		}
+		if finished {
+			break
+		}
+		time.Sleep(5 * time.Second)
+	}
+
+	wg.Wait()
+}
+
+func (a *Append) batchGC(startTS int64, endTS int64, batchSize int, l0Trigger int, getL0Num func() (int, error)) (int64, bool, error) {
+
+	batch := new(leveldb.Batch)
 	deleteNum := 0
+	alreadyGcTS := startTS
 
 	irange := &util.Range{
```

Comment:
maybe we also should avoid this issue https://github.com/pingcap/tidb-binlog/pull/758